src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/MsgPackTraceEventSession.cs

```diff
@@ -0,0 +1,166 @@
+﻿using Microsoft.Diagnostics.Tracing;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+using MsgPack.Serialization;
+using Microsoft.Diagnostics.EventFlow.Utilities.Etw;
+
+namespace Microsoft.Diagnostics.EventFlow.Inputs
+{
+    internal class MsgPackTraceEventSession : StandardTraceEventSession
+    {
+
+        /// <summary>
+        /// MsgPackTraceEventSession allows to decode MsgPack-encoded binary events.
+        /// </summary>
+        /// <param name="sessionNamePrefix">See @base</param>
+        /// <param name="cleanupOldSessions">See @base</param>
+        /// <param name="reuseExisting">See @base</param>
+        /// <param name="healthReporter">See @base</param>
+        public MsgPackTraceEventSession(string sessionNamePrefix, bool cleanupOldSessions, bool reuseExisting, IHealthReporter healthReporter) :
+            base(sessionNamePrefix, cleanupOldSessions, reuseExisting, healthReporter)
+        {
+        }
+
+        /// <summary>
+        /// Convert TraceEvent in MsgPack format to EventData
+        /// </summary>
+        /// <param name="data"></param>
+        /// <param name="reporter"></param>
+        /// <returns></returns>
+        internal static EventData DecodeMessagePack(TraceEvent data, IHealthReporter reporter)
+        {
+            EventData eventData = new EventData
+            {
+                ProviderName = data.ProviderName,
+                Timestamp = data.TimeStamp,
+                Level = (LogLevel)data.Level,
+                Keywords = (long)data.Keywords,
+            };
+
+            if (data.EventDataLength==0)
+            {
+                reporter.ReportProblem("MsgPack decoding failed: empty payload!");
+                return eventData;
+            }
+
+            byte[] managedArray = new byte[data.EventDataLength];
```

Comment:
Will this buffer be allocated every time a new event comes? Can/should we reduce allocation by using something like `ArrayPool<T>`?

Reply:
`Span<T>` will probably be even better if `EventDataLength` is not unreasonably large and if `UnpackMessagePackObject` can work with spans (and if the latter is not true, we might consider contributing a patch ☺️ )